lib/puma/cluster.rb

```diff
@@ -37,7 +37,14 @@ def stop_workers
       @workers.each { |x| x.term }
 
       begin
-        @workers.each { |w| Process.waitpid(w.pid) }
+        pids = @workers.map(&:pid)
+        until pids.empty?
+          if pid = Process.waitpid(-1, Process::WNOHANG)
+            pids.delete(pid)
+            sleep 0.5
+            break
+          end
+        end
```

Comment:
In fact, there are two cycles.

It can be done in one with `redo`. And without the second array.